src/components/shapes/draw.js

```diff
@@ -579,6 +599,224 @@ function setupDragElement(gd, shapePath, shapeOptions, index, shapeLayer, editHe
+    var textangle = options.label.textangle;
+    if(textangle === 'auto') {
+        if(options.type === 'line') {
+            // Auto angle for line is same angle as line
+            textangle = calcTextAngle(shapex0, shapey0, shapex1, shapey1);
+        } else {
+            // Auto angle for all other shapes is 0
+            // The default angle for other shapes is 0 anyway, so there's no reason to
+            // explicitly set the angle to `auto`, this is just covering an edge case
+            textangle = 0;
+        }
+    }
+
+    // Do a fake render so we can get the text bounding box height
+    var _labelText = labelGroup.append('text')
+        .attr(labelTextAttrs)
+        .classed('shape-label-text', true)
+        .text(text);
+    _labelText.call(function(s) {
+        s.call(Drawing.font, font).attr({});
+        svgTextUtils.convertToTspans(s, gd);
+        return s;
+    });
+    var textBB = Drawing.bBox(_labelText.node());
+    _labelText.remove();
+
+    // Calculate correct (x,y) for text
+    // We also determine true xanchor since xanchor depends on position when set to 'auto'
+    var textPos = calcTextPosition(shapex0, shapey0, shapex1, shapey1, options, textangle, textBB);
+    var textx = textPos.textx;
+    var texty = textPos.texty;
+    var xanchor = textPos.xanchor;
+
+    function textLayout(s) {
+        s.call(Drawing.font, font)
+            .attr({
+                'text-anchor': {
+                    left: 'start',
+                    center: 'middle',
+                    right: 'end'
+                }[xanchor],
+                'y': texty,
+                'x': textx,
+                'transform': 'rotate(' + textangle + ',' + textx + ',' + texty + ')'
+            });
+        svgTextUtils.convertToTspans(s, gd);
+        return s;
+    }
+
+    labelText.call(textLayout);
+}
+
+function calcTextAngle(shapex0, shapey0, shapex1, shapey1) {
+    var dy, dx;
+    dx = Math.abs(shapex1 - shapex0);
+    if(shapex1 >= shapex0) {
+        dy = shapey0 - shapey1;
+    } else {
+        dy = shapey1 - shapey0;
+    }
+    return -180 / Math.PI * Math.atan2(dy, dx);
+}
+
+function calcTextPosition(shapex0, shapey0, shapex1, shapey1, shapeOptions, actualTextAngle, textBB) {
+    var textPosition = shapeOptions.label.textposition;
+    var textPadding = shapeOptions.label.padding;
+    var shapeType = shapeOptions.type;
+    var textAngleRad = Math.PI / 180 * actualTextAngle;
+    var xanchor = shapeOptions.label.xanchor;
+    var yanchor = shapeOptions.label.yanchor;
+
+    var textx, texty, paddingX, paddingY;
+
+    // Text position functions differently for lines vs. other shapes
+    if(shapeType === 'line') {
+        // Set base position for start vs. center vs. end of line (default is 'center')
+        if(textPosition.indexOf('start') !== -1) {
+            textx = shapex0;
+            texty = shapey0;
+            if(xanchor === 'auto') xanchor = (shapex1 >= shapex0) ? 'right' : 'left';
```

Comment:
Possibly rewrite as
```suggestion
            if(xanchor === 'auto') xanchor = (shapex1 < shapex0) ? 'left' : 'right';
```
i.e. to avoid `>=` operator.
BTW shouldn't we set to *center* when they are equal?

Reply:
@archmoj yes! Good catch